### ISSUE TITLE
template props intellisense respect to jsdoc type

### ISSUE THREAD

**Issue by bingzheyuan** [vuejs/vetur#1337]
- [x] I have searched through existing issues
- [x] I have read through [docs](https://vuejs.github.io/vetur)
- [x] I have read [FAQ](https://github.com/vuejs/vetur/blob/master/docs/FAQ.md)

## Info

- Platform: win7 x64
- Vetur version: 0.21.0
- VS Code version: 1.35.1

## Problem
![image](https://user-images.githubusercontent.com/5099742/59590460-6dc7c100-90ec-11e9-9705-db165c589aa5.png)
Props intelliSense has been provided to template interpolation since #1083 , however for `Object` type seems not enough, is that possible to provide intelliSence type base on jsDoc `@type`? For the code in image, when I type `.` after `sum`, there's no intelliSence at all, I shall expect to select from `v1` and `v2`.

## Reproducible Case
Create a Vue component with following code 
```vue
// Test.vue
<template>
  <div class="container">
    <p v-if="loading">loading...{{sum}}</p>
    <h1 v-if="!loading">
      <p>{{sum.v1}}</p> <!--  <--- here after sum I shall have intelliSense -->
      <p>{{sum.v2}}</p>
    </h1>
  </div>
</template>

<script>
export default {
  name: 'Test',
  props: {
    loading: {
      type: Boolean,
      required: true
    },
    /** @type {{v1: number, v2: number}} */
    sum: {
      type: Object,
      required: true
    }
  },
  data() {
    return {
      message: 'Using Parcel In A Vue.js App',
    };
  },
};
</script>
```

**Comment by azuryus** [comment]
You can declare your props' types like this:
```js
export default {
  props: {
    /** @type {import('vue').PropOptions<{v1: number, v2: number}>} */
    sum: {
      type: Object,
      required: true
    }
  }
}
```

I don't know if it's the recommended way, but it will enable IntelliSense with your object's properties.

**Comment by bingzheyuan** [comment]
@azuryus  Thanks for the hint.

**Comment by octref** [comment]
Vetur currently walks through AST to find out the information. For doing this, it needs to get the type information for each props.

**Comment by yoyo930021** [comment]
If you use typescript:
```typescript
import { PropType } from 'vue'

export default {
  props: {
    sum: {
      type: Object as PropType<{v1: number, v2: number}>,
      required: true
    }
  }
}
```
javascript:
```javascript
export default {
  props: {
    sum: {
      /** @type {import('vue').PropType<{v1: number, v2: number}>} */
      type: Object,
      required: true
    }
  }
}
```

**Comment by xiezht** [comment]
> If you use typescript:
> 
> ```ts
> import { PropType } from 'vue'
> 
> export default {
>   props: {
>     sum: {
>       type: Object as PropType<{v1: number, v2: number}>,
>       required: true
>     }
>   }
> }
> ```
> 
> javascript:
> 
> ```js
> export default {
>   props: {
>     sum: {
>       /** @type {import('vue').PropType<{v1: number, v2: number}>} */
>       type: Object,
>       required: true
>     }
>   }
> }
> ```

Thanks, your answer help me a lot. But i have another problem, how can i use jsdoc for "data" in vue component. 

Something may like this:

```js
...
  mounted() {},
  data() {
    return {
      /**
        * @type {PlanDto}
        */
      planList: [], // 所有的广告计划列表
    };
  },
...
```

**Comment by rchl** [comment]
Yes, like that.

Also, to correct @yoyo930021 [comment](https://github.com/vuejs/vetur/issues/1337#issuecomment-667482279), for javascript you place the JSDoc comment above the prop name and not above the `type`, like so:

```js
export default {
  props: {
    /** @type {import('vue').PropType<{v1: number, v2: number}>} */
    sum: {
      type: Object,
      required: true
    }
  }
}
```

**Comment by xiezht** [comment]
@rchl It works when i hover on the variable, like this:

![image](https://user-images.githubusercontent.com/18051618/89625001-a8c91d00-d8c9-11ea-8db8-a8869c2044f4.png)

But it maybe not the perfect result. What i hope is that it can perform like JSDOC for vue's props. Like this:

![image](https://user-images.githubusercontent.com/18051618/89625267-31e05400-d8ca-11ea-9a29-e82a2afa001c.png)

Is it possible ?

**Comment by rchl** [comment]
Yes, it should work like that for data also. If it doesn't then you need to provide more information on what you are doing. Ideally, a small repo that reproduces.

Note that if you are missing some annotations in your code, it might break types in general. For example, you should add annotations to your methods too. And computed properties. At least when they are non-trivial or contain arguments.

**Comment by xiezht** [comment]
Oh, i got it. The jsdoc comment should be about the "data" name, instead of in the object. Like this:

![image](https://user-images.githubusercontent.com/18051618/89629303-2db73500-d8d0-11ea-8689-aa3bebb408e4.png)

Then the property of data object can be auto-completed.

![image](https://user-images.githubusercontent.com/18051618/89629267-1d06bf00-d8d0-11ea-86a0-f7f3ffb42228.png)

Thanks for help. @rchl

**Comment by rchl** [comment]
If it works for you then go ahead but I'm always annotating individual data properties instead and it works for me that way.

**Comment by octref** [comment]
Related to #2193.

**Comment by octref** [comment]
![image](https://user-images.githubusercontent.com/4033249/91250871-a0f3ee80-e78c-11ea-8f76-d66b5b0539e2.png)

**Comment by fortune-max** [comment]
> javascript:
> 
> ```js
> export default {
>   props: {
>     sum: {
>       /** @type {import('vue').PropType<{v1: number, v2: number}>} */
>       type: Object,
>       required: true
>     }
>   }
> }
> ```

As mentioned by [yoyo930021](https://github.com/yoyo930021), this is indeed the right way (tested with Vue 3.2.47, jsdoc).

But I noticed my prop types were still referred to with `any` within my computed/methods declarations until I added a data declaration (even if empty) for the types to be recognized within computed/methods/etc.

```
data() { }
```

Easy to miss if you only use props in your component, but no data field declarations.

[I use Volar so not sure if Volar specific, but this seems to be the most likely page anyone searching on Google may stumble upon, and I'd rather no one spend as much time as I did figuring it out]
